irukandji epic (#135)
ghga-de/epic-docs: 66-irukandji/technical_specification.md

# Schemapack (Irukandji)
**Epic Type:** Implementation Epic

Epic planning and implementation follow the
[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).

## Scope

### Outline:
This epic aims to make datapack navigation independent of its corresponding schemapack. Currently, since the relation class is not explicitly provided in the datapack, resolving target classes requires referring back to the schemapack. This epic introduces changes to eliminate this dependency, allowing the datapack to be self-contained for navigation.

Additionally, this epic introduces configurable embedding depth in datapack denormalization. This ensures more control over how deeply related entities are embedded, enabling partial embedding instead of always including all related data at the highest level.

### Included/Required:

#### Example datapacks/schemapacks

> 1. Update all the examples so that the datapack also has the rootClass, targetResources, rootResource and rootClass if it is specified in schemapack. 

Reviewer: ```suggestion
Update all examples so that the datapack also contains the rootClass, targetResources, rootResource and rootClass if it is specified in schemapack. 
```

#### Datapack specification

> 2. Change the datapack specification 

Reviewer: ```suggestion
1. Change the datapack specification in the following way:
```

   1. if there is  root resource, root resource has to exits

>    2. validating against datapack specification will rely on pydantic to ensure the root resource check. 

Reviewer: Root resource for this check has has to exist where, schemapack?
Or is this more of a "If a root resource is required/should be specified, a concrete instance has to exist"?

Please elaborate on how validation is handled if this is optional as long as the datapack is not rooted, e.g. custom validators?

Author: Schemapack includes two types of validation. One verifies whether a schemapack or datapack conforms to its specifications. The plug-ins, on the other hand, validate whether data provided as a datapack aligns with a given model, which is a schemapack.

Everything discussed here pertains solely to the datapack specifications. So if a resource is given as a root resource in a datapack, that resource cannot be a random string, but it has to exist among the data.

Author: rephrased.

> 3. The datapack specification will check if a datapack is valid or not. The validation plugins check if datapack is valid given a schemapack. 

Reviewer: ```suggestion
2. The datapack specification will check if a datapack is valid or not. The validation plugins check if datapack is valid given a schemapack. 
```

Not sure about the phrasing here, are those statements slightly redundant or does this mean that a datapack is checked for internal consistency first and then it is validated against a given schemapack?

Author: Yes. Rephrased

   1. shift the functionality of `unknown_root_resource.py` to datapack specification. 
   2. shift the functionality of `target_id.py` to datapack specification.

#### Validatio Plugings

> 1. The following validators require are required to remain unchanged:

Reviewer: ```suggestion
1. The following validators will remain unchanged:
```

   1. `content_schema.py`
   2. `missing_class.py`
   3. `missing_origin.py`
   4. `missing_relations.py`
   5. `missing_target.py`
   6. `multiple_target.py`
   7. `one_to_many_overlap.py`
   8. `unknown_class.py`
   9. `unknown_relations.py`
2.  The following validators are obsolete and can be deleted:

>     1. `unknown_root_resource.py`, since the functionality is shifted to datapack specification. 

Reviewer: ```suggestion
    1. `unknown_root_resource.py`, since the functionality is shifted to the datapack specification. 
```

>     2. `target_id.py`, since the functionality is shifted to datapack specification. 

Reviewer: ```suggestion
    2. `target_id.py`, since the functionality is shifted to the datapack specification. 
```

3. If schemapack has a root class defined, `expected_root` plugin checks that the datapack has a root resource. 
   1. Extend `expected_root.py`, so that it checks if the datapack also has a root class. 
   2. Extend `expected_root`, so that it also checks schemapack has a root class defined, if datapack has a root class and a root resource. 

> 4. If schemapack does not have a root class, `unexpected_root` checks that datapack has not a root source. 

Reviewer: ```suggestion
4. If schemapack does not have a root class, `unexpected_root` checks that datapack does not have a root resource. 
```

   1. Extend it, so that it checks that datapack does not have a root class defined either. 

> 5. Merge `expected_root` and `unexpected_root` plugins into one root validator named `root_duality.py`

Reviewer: If this should be done from the get go, why not describe this at the top and then list what each subplugin is supposed to do instead?

Author: I think this ordering fits more to the order of the tasks.

> 6. `root_duality` must check the consistency

Reviewer: ```suggestion
6. `root_duality` must check for consistency between schemapack and datapack, i.e. that
```

>    1. root class is same between the schemapack and the datapack

Reviewer: ```suggestion
   1. the root class is same between the schemapack and the datapack
```

>    2. if root class in None in schemapack, it must be None in datapack.

Reviewer: ```suggestion
   2. if the root class in None in schemapack, it must be None in datapack.
```

#### Embedded Profiles

The denormalization process currently embeds all related entities into a single JSON output at the highest level. A configurable denormalization depth to allow more control over how deeply related entities are embedded will be implemented.

> For example, given that an experiment has relations to sample, sample has relations to files, the current implementation of denormalization embeds everything into the experiment including the complete information of the files referred by the samples. However, implementing an embedding depth will enable the denormalization of the experiment that embed samples but will keep the referred file information unchanged/not embedded. 

Reviewer: Could use a bit more information on how the embedding depth will be implemented, i.e. is this a simple integer value or list of classes that should not be embedded?

Author: It should support the level of nestedness for sure. I'm not sure if we need to configure which classes should not be embedded, but I doubt it will be necessary.

Reviewer 2: We should make that decision now and document it in the epic spec, at leasts a preliminary one

Author: But this section was already extended with a paragraph. I don't know why it is not visible here.
